### PR TITLE
MULE-15473: Added new provider interface for dynamic TLS configuration.

### PR DESCRIPTION
**Pull request by 3miliano**

### CODE REVIEW

File: src/main/java/org/mule/runtime/api/tls/DynamicTlsProvider.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ * a connector that already has TLS enabled it will override the previous TLS context factory without the possiblity
+ * to rollback unless you have a reference to it.
+ *
+ * @since 4.2
+ */
+@NoImplement
+public interface DynamicTlsProvider {
+
+  /**
+   * Enable TLS dynamically on this server using the supplied TLS context factory
```
Comment (afelisatti) [inline src/main/java/org/mule/runtime/api/tls/DynamicTlsProvider.java]: Will this be limited to servers? Seems broad enough to be used in any component that features TLS.

File: src/main/java/org/mule/runtime/api/tls/DynamicTlsProvider.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+   *
+   * @param tlsContextFactory The TLS context factory to be used when dynamically enabled TLS at this server
+   */
+  void enableTls(TlsContextFactory tlsContextFactory);
+
+  /**
+   * Disable TLS dynamically. It should work for when the TLS was dynamically enabled or even when the TLS
+   * was statically configured in the listener configuration section of the application.
```
Comment (afelisatti) [inline src/main/java/org/mule/runtime/api/tls/DynamicTlsProvider.java]: Same as before.

File: src/main/java/org/mule/runtime/api/tls/DynamicTlsProvider.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ * Connectors that implement this interface allow to enable/disable TLS on-the-fly using the provided TLS context
+ * factory. Only one TLS context factory is allowed to be configured at any one time. If you re-enable TLS for
+ * a connector that already has TLS enabled it will override the previous TLS context factory without the possiblity
+ * to rollback unless you have a reference to it.
+ *
+ * @since 4.2
+ */
+@NoImplement
```
[the remark this answers is not in the document]
Comment (3miliano) [reply src/main/java/org/mule/runtime/api/tls/DynamicTlsProvider.java]: That was a copy and paste error. I'll remove it.

File: src/main/java/org/mule/runtime/api/tls/DynamicTlsProvider.java (hunk 1 above)
Comment (3miliano) [reply src/main/java/org/mule/runtime/api/tls/DynamicTlsProvider.java]: No it won't.

File: src/main/java/org/mule/runtime/api/tls/DynamicTlsProvider.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.api.tls;
+
+import org.mule.api.annotation.NoImplement;
+
+/**
+ * Connectors that implement this interface allow to enable/disable TLS on-the-fly using the provided TLS context
```
[the remark this answers is not in the document]
Comment (3miliano) [reply src/main/java/org/mule/runtime/api/tls/DynamicTlsProvider.java]: Elaborate, how come it won't be limited to those? Where does TLS applies that isn't a connector?

[the remark this answers is not in the document]
Comment (afelisatti) [reply src/main/java/org/mule/runtime/api/tls/DynamicTlsProvider.java]: The service we are adding it to is not technically a connector but something they use internally. In fact it's very likely that another interface is used whenever this feature is added to the extensions framework.

**Pull request closed without merging**